File: app/src/main/java/com/fastaccess/ui/widgets/recyclerview/scroll/RecyclerViewFastScroller.java
package com.fastaccess.ui.widgets.recyclerview.scroll;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.design.widget.AppBarLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;

import com.fastaccess.R;
import com.fastaccess.helper.ActivityHelper;
import com.fastaccess.helper.Logger;

import it.sephiroth.android.library.bottomnavigation.BottomNavigation;

public class RecyclerViewFastScroller extends FrameLayout {

    private static final int TRACK_SNAP_RANGE = 5;
    private ImageView scrollerView;
    private ImageButton scrollTop;
    private int height;
    private RecyclerView recyclerView;
    private RecyclerView.LayoutManager layoutManager;
    private AppBarLayout appBarLayout;
    private BottomNavigation bottomNavigation;
    private boolean toggled;
    private boolean registeredObserver = false;

    public RecyclerViewFastScroller(Context context) {
        super(context);
        init();
    }

    public RecyclerViewFastScroller(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RecyclerViewFastScroller(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @Override protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        height = h;
    }

    @SuppressLint("ClickableViewAccessibility") @Override public boolean onTouchEvent(@NonNull MotionEvent event) {
        int action = event.getAction();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                if (event.getX() < (scrollerView.getX() - scrollerView.getPaddingStart())) return false;
                scrollerView.setSelected(true);
                hideAppbar();
            case MotionEvent.ACTION_MOVE:
                float y = event.getY();
                setScrollerHeight(y);
                setRecyclerViewPosition(y);
                return true;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                scrollerView.setSelected(false);
                showAppbar();
                return true;
        }
        return super.onTouchEvent(event);
    }

    @Override protected void onDetachedFromWindow() {
        if (recyclerView != null) {
            recyclerView.removeOnScrollListener(onScrollListener);
            safelyUnregisterObserver();
        }
        appBarLayout = null;
        bottomNavigation = null;
        super.onDetachedFromWindow();
    }

    private void safelyUnregisterObserver() {
        try {// rare case
            if (registeredObserver && recyclerView.getAdapter() != null) {
                recyclerView.getAdapter().unregisterAdapterDataObserver(observer);
            }
        } catch (Exception ignored) {}
    }

    protected void init() {
        setVisibility(GONE);
        setClipChildren(false);
        LayoutInflater inflater = LayoutInflater.from(getContext());
        inflater.inflate(R.layout.fastscroller_layout, this);
        scrollerView = findViewById(R.id.fast_scroller_handle);
        setVisibility(VISIBLE);
        Activity activity = ActivityHelper.getActivity(getContext());
        if (activity != null) {
            appBarLayout = activity.findViewById(R.id.appbar);
            bottomNavigation = activity.findViewById(R.id.bottomNavigation);
        }
    }

    protected void hideAppbar() {
        if (!toggled) {
            if (appBarLayout != null) {
                appBarLayout.setExpanded(false, true);
            }
            if (bottomNavigation != null) {
                bottomNavigation.setExpanded(false, true);
            }
            toggled = true;
        }
    }

    protected void showAppbar() {
        if (toggled) {
            if (scrollerView.getY() == 0) {
                if (appBarLayout != null) {
                    appBarLayout.setExpanded(true, true);
                }
                if (bottomNavigation != null) {
                    bottomNavigation.setExpanded(true, true);
                }
                toggled = false;
            }
        }
    }

    public void attachRecyclerView(final RecyclerView recyclerView) {
        if (this.recyclerView == null) {
            this.recyclerView = recyclerView;
            this.layoutManager = recyclerView.getLayoutManager();
            this.recyclerView.addOnScrollListener(onScrollListener);
            if (recyclerView.getAdapter() != null && !registeredObserver) {
                recyclerView.getAdapter().registerAdapterDataObserver(observer);
                registeredObserver = true;
            }
            hideShow();
            initScrollHeight();
        }
    }

    private void initScrollHeight() {
        if (recyclerView.computeVerticalScrollOffset() == 0) {
            this.recyclerView.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
                @Override public boolean onPreDraw() {
                    RecyclerViewFastScroller.this.recyclerView.getViewTreeObserver().removeOnPreDrawListener(this);
                    iniHeight();
                    return true;
                }
            });
        } else {
            iniHeight();
        }
    }

    protected void iniHeight() {
        if (scrollerView.isSelected()) return;
        int verticalScrollOffset = RecyclerViewFastScroller.this.recyclerView.computeVerticalScrollOffset();
        int verticalScrollRange = RecyclerViewFastScroller.this.computeVerticalScrollRange();
        float proportion = (float) verticalScrollOffset / ((float) verticalScrollRange - height);
        setScrollerHeight(height * proportion);
    }

    private void setRecyclerViewPosition(float y) {
        Logger.e(y);
        if (recyclerView != null) {
            int itemCount = recyclerView.getAdapter().getItemCount();
            float proportion;
            if (scrollerView.getY() == 0) {
                proportion = 0f;
            } else if (scrollerView.getY() + scrollerView.getHeight() >= height - TRACK_SNAP_RANGE) {
                proportion = 1f;
            } else {
                proportion = y / (float) height;
            }
            int targetPos = getValueInRange(itemCount - 1, (int) (proportion * (float) itemCount));
            if (layoutManager instanceof StaggeredGridLayoutManager) {
                ((StaggeredGridLayoutManager) layoutManager).scrollToPositionWithOffset(targetPos, 0);
            } else if (layoutManager instanceof GridLayoutManager) {
                ((GridLayoutManager) layoutManager).scrollToPositionWithOffset(targetPos, 0);
            } else {
                ((LinearLayoutManager) layoutManager).scrollToPositionWithOffset(targetPos, 0);
            }
        }
    }

    private static int getValueInRange(int max, int value) {
        return Math.min(Math.max(0, value), max);
    }

    private void setScrollerHeight(float y) {
        int handleHeight = scrollerView.getHeight();
        scrollerView.setY(getValueInRange(height - handleHeight, (int) (y - handleHeight / 2)));
    }

    private final RecyclerView.OnScrollListener onScrollListener = new RecyclerView.OnScrollListener() {
        @Override public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            if (scrollerView.isSelected()) return;
            int verticalScrollOffset = recyclerView.computeVerticalScrollOffset();
            int verticalScrollRange = recyclerView.computeVerticalScrollRange();
            float proportion = (float) verticalScrollOffset / ((float) verticalScrollRange - height);
            setScrollerHeight(height * proportion);
        }
    };

    private final RecyclerView.AdapterDataObserver observer = new RecyclerView.AdapterDataObserver() {
        @Override public void onItemRangeInserted(int positionStart, int itemCount) {
            super.onItemRangeInserted(positionStart, itemCount);
            hideShow();
        }

        @Override public void onItemRangeRemoved(int positionStart, int itemCount) {
            super.onItemRangeRemoved(positionStart, itemCount);
            hideShow();
        }

        @Override public void onChanged() {
            super.onChanged();
            hideShow();

        }
    };

    protected void hideShow() {
        if (recyclerView != null && recyclerView.getAdapter() != null) {
            setVisibility(recyclerView.getAdapter().getItemCount() > 10 ? VISIBLE : GONE);
        } else {
            setVisibility(GONE);
        }
    }
}

File: app/src/main/java/com/fastaccess/github/ui/widget/recyclerview/RecyclerViewFastScroller.kt
package com.fastaccess.github.ui.widget.recyclerview

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewTreeObserver
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.fastaccess.github.R
import com.google.android.material.appbar.AppBarLayout

open class RecyclerViewFastScroller : FrameLayout {

    private lateinit var scrollerView: ImageView
    private var scrollerHeight: Int = 0
    private var recyclerView: RecyclerView? = null
    private var layoutManager: RecyclerView.LayoutManager? = null
    private var appBarLayout: AppBarLayout? = null
    private var toggled: Boolean = false

    private val onScrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            if (scrollerView.isSelected) return
            val verticalScrollOffset = recyclerView.computeVerticalScrollOffset()
            val verticalScrollRange = recyclerView.computeVerticalScrollRange()
            val proportion = verticalScrollOffset.toFloat() / (verticalScrollRange.toFloat() - scrollerHeight)
            setScrollerHeight(scrollerHeight * proportion)
        }
    }

    private val observer = object : RecyclerView.AdapterDataObserver() {
        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            super.onItemRangeInserted(positionStart, itemCount)
            hideShow()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            super.onItemRangeRemoved(positionStart, itemCount)
            hideShow()
        }

        override fun onChanged() {
            super.onChanged()
            hideShow()

        }
    }

    constructor(context: Context) : super(context) {
        init()
    }

    @JvmOverloads constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int = 0) : super(context, attrs, defStyleAttr) {
        init()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        scrollerHeight = h
    }

    @SuppressLint("ClickableViewAccessibility") override fun onTouchEvent(event: MotionEvent): Boolean {
        val action = event.action
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                if (event.x < scrollerView.x - scrollerView.paddingStart) return false
                scrollerView.isSelected = true
                hideAppbar()
                val y = event.y
                setScrollerHeight(y)
                setRecyclerViewPosition(y)
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                val y = event.y
                setScrollerHeight(y)
                setRecyclerViewPosition(y)
                return true
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                scrollerView.isSelected = false
                showAppbar()
                return true
            }
        }
        return super.onTouchEvent(event)
    }

    override fun onDetachedFromWindow() {
        recyclerView?.removeOnScrollListener(onScrollListener)
        safelyUnregisterObserver()
        appBarLayout = null
        super.onDetachedFromWindow()
    }

    private fun safelyUnregisterObserver() {
        kotlin.runCatching {
            recyclerView?.adapter?.unregisterAdapterDataObserver(observer)
        }
    }

    private fun init() {
        visibility = View.GONE
        clipChildren = false
        val inflater = LayoutInflater.from(context)
        inflater.inflate(R.layout.fastscroller_layout, this)
        scrollerView = findViewById(R.id.fast_scroller_handle)
        visibility = View.VISIBLE
    }

    private fun hideAppbar() {
        if (!toggled) {
            appBarLayout?.setExpanded(false, true)
            toggled = true
        }
    }

    private fun showAppbar() {
        if (toggled) {
            if (scrollerView.y == 0f) {
                appBarLayout?.setExpanded(true, true)
                toggled = false
            }
        }
    }

    fun attachRecyclerView(recyclerView: RecyclerView, appBarLayout: AppBarLayout? = null) {
        this.appBarLayout = appBarLayout
        if (this.recyclerView == null) {
            this.recyclerView = recyclerView
            this.layoutManager = recyclerView.layoutManager
            this.recyclerView?.addOnScrollListener(onScrollListener)
            if (recyclerView.adapter != null) {
                runCatching { recyclerView.adapter?.registerAdapterDataObserver(observer) }
            }
            hideShow()
            initScrollHeight()
        }
    }

    private fun initScrollHeight() {
        if (recyclerView?.computeVerticalScrollOffset() == 0) {
            this.recyclerView?.viewTreeObserver?.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
                override fun onPreDraw(): Boolean {
                    this@RecyclerViewFastScroller.recyclerView?.viewTreeObserver?.removeOnPreDrawListener(this)
                    iniHeight()
                    return true
                }
            })
        } else {
            iniHeight()
        }
    }

    protected fun iniHeight() {
        if (scrollerView.isSelected) return
        val verticalScrollOffset = this@RecyclerViewFastScroller.recyclerView?.computeVerticalScrollOffset() ?: 0
        val verticalScrollRange = this@RecyclerViewFastScroller.computeVerticalScrollRange()
        val proportion = verticalScrollOffset.toFloat() / (verticalScrollRange.toFloat() - scrollerHeight)
        setScrollerHeight(scrollerHeight * proportion)
    }

    private fun setRecyclerViewPosition(y: Float) {
        if (recyclerView != null) {
            val itemCount = recyclerView?.adapter?.itemCount ?: 0
            val proportion: Float = when {
                scrollerView.y == 0f -> 0f
                scrollerView.y + scrollerView.height >= scrollerHeight - TRACK_SNAP_RANGE -> 1f
                else -> y / scrollerHeight.toFloat()
            }
            val targetPos = getValueInRange(itemCount - 1, (proportion * itemCount).toInt())
            when (layoutManager) {
                is StaggeredGridLayoutManager -> (layoutManager as StaggeredGridLayoutManager).scrollToPositionWithOffset(targetPos, 0)
                is GridLayoutManager -> (layoutManager as GridLayoutManager).scrollToPositionWithOffset(targetPos, 0)
                else -> (layoutManager as LinearLayoutManager).scrollToPositionWithOffset(targetPos, 0)
            }
        }
    }

    private fun setScrollerHeight(y: Float) {
        val handleHeight = scrollerView.height
        scrollerView.y = getValueInRange(scrollerHeight - handleHeight, (y - handleHeight / 2).toInt()).toFloat()
    }

    protected fun hideShow() {
        visibility = if (recyclerView != null && recyclerView?.adapter != null) {
            if (recyclerView?.adapter?.itemCount ?: 0 > 10) View.VISIBLE else View.GONE
        } else {
            View.GONE
        }
    }

    companion object {
        private const val TRACK_SNAP_RANGE = 5
        private fun getValueInRange(max: Int, value: Int): Int {
            return Math.min(Math.max(0, value), max)
        }
    }
}